dynawo/sources/Solvers/AlgebraicSolvers/DYNSolverKINAlgRestoration.cpp

```diff
@@ -0,0 +1,381 @@
+//
+// Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
+// See AUTHORS.txt
+// All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+// simulation tool for power systems.
+//
+
+/**
+ * @file  DYNSolverKINAlgRestoration.cpp
+ *
+ * @brief Implementation of solver based on sundials/KINSOL solver
+ *
+ */
+
+
+#include <kinsol/kinsol.h>
+#include <sunlinsol/sunlinsol_klu.h>
+#include <sundials/sundials_types.h>
+#include <sundials/sundials_math.h>
+#include <sundials/sundials_sparse.h>
+#include <nvector/nvector_serial.h>
+#include <string.h>
+#include <vector>
+#include <cmath>
+#include <map>
+#include <algorithm>
+#include <iomanip>
+
+#include "DYNSolverKINAlgRestoration.h"
+#include "DYNSolverCommon.h"
+#include "DYNTrace.h"
+#include "DYNMacrosMessage.h"
+#include "DYNModel.h"
+#include "DYNSparseMatrix.h"
+
+using std::vector;
+using std::map;
+using std::string;
+using std::stringstream;
+using boost::shared_ptr;
+
+namespace DYN {
+
+SolverKINAlgRestoration::SolverKINAlgRestoration() :
+model_(),
+mode_(KIN_NORMAL) {
+  SolverKINCommon();
+}
+
+SolverKINAlgRestoration::~SolverKINAlgRestoration() {
+  clean();
+}
+
+void
+SolverKINAlgRestoration::init(const shared_ptr<Model>& model, modeKin_t mode, double fnormtol, double scsteptol,
+                              double mxnewtstep, int msbset, int mxiter, int printfl) {
+  // (1) Arguments
+  // --------------
+  clean();
+
+  model_ = model;
+  mode_ = mode;
+
+  // (2) Size of the problem
+  // -------------------------------
+  fType_.resize(model->sizeF());
+  std::copy(model_->getFType(), model_->getFType() + model->sizeF(), fType_.begin());
+
+  vId_.resize(model->sizeF());
+  std::copy(model_->getYType(), model_->getYType() + model->sizeF(), vId_.begin());
```

Comment:
why sizeF and not sizeY? even if they are supposed to be equal, I'm not sure std::copy makes a size check and we might have bad memory writing if something goes wrong

Reply:
Right, copy/paste from previous code but it was dangerous (and certainly a copy/paste error ;))